File: legacy/python_engine/core/engine/battle.py
"""アタック宣言・バトル解決・勝敗判定（GameManager からの移管・第1引数 gm）。"""
from __future__ import annotations

from ..journal import JournaledDict, JournaledList
from ...models.enums import CardType, TriggerType, Zone, ActionType, Phase


def declare_attack(gm, attacker: Card, target: Card):
    attacker_owner, _ = gm._find_card_location(attacker)
    target_owner, _ = gm._find_card_location(target)
    gm._validate_action(attacker_owner, "MAIN_ACTION")
    # 先攻・後攻ともに「自分の最初のターン」はリーダー・キャラのいずれもアタックできない（公式準拠）。
    # ターンは先攻=turn_count 1、後攻=turn_count 2 と交互に進むため、turn_count <= 2 が
    # 両プレイヤーの最初のターンを覆う。
    if gm.turn_count <= 2:
        raise ValueError("最初のターンはアタックできません。")
    if "ATTACK_DISABLE" in attacker.flags or "ATTACK_DISABLE" in attacker.timed_flags: raise ValueError("このカードは効果によりアタックできません。")
    if "CANNOT_REST" in attacker.timed_flags: raise ValueError("このカードは効果によりレストにできないためアタックできません。")
    if attacker.is_rest: raise ValueError("アタックするカードはアクティブ状態でなければなりません。")
    # 召喚酔い: 登場したターンのキャラは攻撃できない。ただし「速攻」を持てば可。
    # リーダーは is_newly_played=False のため影響を受けない。
    if (attacker.master.type == CardType.CHARACTER
            and attacker.is_newly_played
            and not attacker.has_keyword("速攻")):
        raise ValueError("登場したターンのキャラクターは攻撃できません（速攻を除く）。")
    # 自己制限（self_cannot）:「リーダーにアタックできない」。相手リーダーへの攻撃宣言を弾く。
    if (target.master.type == CardType.LEADER
            and attacker_owner is not None
            and gm._active_restriction(attacker_owner, "CANNOT_ATTACK_LEADER")):
        raise ValueError("効果により、このターンはリーダーにアタックできません。")
    if (target.master.type == CardType.CHARACTER and not target.is_rest
            and not attacker.has_keyword("ATTACK_ACTIVE")):
        raise ValueError("レスト状態のキャラクターのみ攻撃可能です。")
    # アタック税（OP08-043「アタックする際、自身の手札N枚を捨てなければアタックできない」）。
    # 付与された ATTACK_TAX_DISCARD_N フラグがあれば、手札N枚を支払えるときのみアタック可。
    tax_flags = [f for f in (attacker.flags | attacker.timed_flags)
                 if isinstance(f, str) and f.startswith("ATTACK_TAX_DISCARD_")]
    if tax_flags:
        need = max(int(f.rsplit("_", 1)[1]) for f in tax_flags)
        if len(attacker_owner.hand) < need:
            raise ValueError(f"アタックするには手札{need}枚を捨てる必要があり、手札が足りません。")
        # コスト支払い: 手札N枚を捨てる。どの札を捨てるかは本来プレイヤー選択だが、宣言経路を
        # 中断させないため先頭からN枚を捨てる（捨て札選択の対話化は今後の課題）。
        for _ in range(need):
            attacker_owner.trash.append(attacker_owner.hand.pop(0))
    attacker.is_rest = True
    gm.active_battle = JournaledDict({"attacker": attacker, "target": target, "attacker_owner": attacker_owner, "target_owner": target_owner, "counter_buff": 0})

    # アタック時/相手のアタック時トリガーを順に解決する。途中でいずれかが対象選択や
    # 選択(Choice)で中断した場合、解決前にブロッカー/カウンター段階へ進むと、未解決の
    # interaction とカウンター操作が衝突する（"期待:CHOICE" エラー）。トリガーを待ち行列に
    # 積み、_advance_battle_triggers で1つずつ解決し、全て片付いてからフェイズ遷移する。
    triggers = []
    if attacker.master.abilities:
        for ability in attacker.master.abilities:
            if ability.trigger == TriggerType.ON_ATTACK:
                triggers.append((attacker_owner, ability, attacker))
            # 「このキャラがレストになった時」(ON_REST) はアタック宣言で自身がレストになった
            # 瞬間に誘発する（要因＝アタックなので「効果で」限定の能力は対象外）。
            # OP14-119/027/028/032/035 等。CONTEXT/ターン1回条件は resolve_ability が評価。
            elif (ability.trigger == TriggerType.ON_REST
                  and gm._rest_subject_matches(ability, attacker, attacker,
                                                 attacker_owner, by_attack=True)):
                triggers.append((attacker_owner, ability, attacker))
    opp_cards = ([target_owner.leader] if target_owner.leader else []) + target_owner.field
    for card in opp_cards:
        for ability in card.master.abilities:
            if ability.trigger == TriggerType.ON_OPP_ATTACK:
                triggers.append((target_owner, ability, card))
    gm._battle_triggers = JournaledList(triggers)
    gm._advance_battle_triggers()

def _advance_battle_triggers(gm):
    """積んだバトルトリガーを順に解決し、全て解決後に防御フェイズへ遷移する。
    途中で interaction が立ったら中断（resolve_interaction が解決後に再度呼ぶ）。"""
    if not gm.active_battle:
        gm._battle_triggers = JournaledList()
        return
    while getattr(gm, "_battle_triggers", None):
        player, ability, card = gm._battle_triggers.pop(0)
        gm.resolve_ability(player, ability, source_card=card)
        if gm.active_interaction:
            return  # 中断: 解決後に resolve_interaction から再開される
    # 全トリガー解決 → ブロッカー/カウンター段階へ
    target_owner = gm.active_battle["target_owner"]
    if gm.has_blocker(target_owner):
        gm.phase = Phase.BLOCK_STEP
    else:
        gm.phase = Phase.BATTLE_COUNTER

def handle_block(gm, blocker: Optional[Card] = None):
    if not gm.active_battle: return
    target_owner = gm.active_battle["target_owner"]; gm._validate_action(target_owner, "SELECT_BLOCKER")
    if blocker:
        blocker.is_rest = True
        gm.active_battle["target"] = blocker
        # 【ブロック時】効果を発動する（従来は未発火＝14枚が no-op だった）。
        if blocker.master.abilities and not blocker.is_effect_negated and not blocker.negated:
            for ability in blocker.master.abilities:
                if ability.trigger == TriggerType.ON_BLOCK:
                    gm.resolve_ability(target_owner, ability, source_card=blocker)
        if gm.active_interaction:
            # ブロック時効果が対象選択等で中断した場合はここで返す（resume が継続）。
            return
    gm.phase = Phase.BATTLE_COUNTER;

def apply_counter(gm, player: Player, counter_card: Optional[Card] = None, don_list: Optional[List[DonInstance]] = None):
    if not gm.active_battle: return
    if counter_card is None: gm.resolve_attack(); return
    gm._validate_action(player, "SELECT_COUNTER")
    if counter_card.master.type == CardType.EVENT:
        gm.pay_cost(player, counter_card.master.cost, don_list)
        for ability in counter_card.master.abilities:
            if ability.trigger == TriggerType.COUNTER: gm.resolve_ability(player, ability, source_card=counter_card)
        # 「自分のキャラすべては、このターン中、…代わりに〜できる」(EB02-030) のような
        # 継続付与型の置換を登録する。イベントは即トラッシュで場に残らないため、
        # _find_replacement の場上 protector 走査では拾えない。player へ this-turn 付与する。
        gm._register_granted_replacements(player, counter_card)
        gm.move_card(counter_card, Zone.TRASH, player)
    else:
        counter_value = getattr(counter_card, "current_counter", counter_card.master.counter or 0); gm.active_battle["counter_buff"] += counter_value
        gm.move_card(counter_card, Zone.TRASH, player)

def resolve_attack(gm):
    if not gm.active_battle: return
    attacker = gm.active_battle["attacker"]; target = gm.active_battle["target"]
    attacker_owner = gm.active_battle["attacker_owner"]; target_owner = gm.active_battle["target_owner"]
    counter_buff = gm.active_battle.get("counter_buff", 0)
    is_my_turn = (attacker_owner == gm.turn_player); is_target_turn = (target_owner == gm.turn_player)
    attacker_pwr = attacker.get_power(is_my_turn); target_pwr = target.get_power(is_target_turn) + counter_buff
    life_lost = 0
    if target == target_owner.leader:
        if attacker_pwr >= target_pwr:
            damage_amount = 2 if attacker.has_keyword("ダブルアタック") else 1; is_banish = attacker.has_keyword("バニッシュ")
            for _ in range(damage_amount):
                if target_owner.life:
                    life_card = target_owner.life.pop(0)
                    dest_zone = Zone.TRASH if is_banish else Zone.HAND
                    trigger_ability = None if is_banish else next(
                        (a for a in life_card.master.abilities if a.trigger == TriggerType.TRIGGER), None
                    )
                    gm.move_card(life_card, dest_zone, target_owner)
                    life_lost += 1
                    # 【トリガー】は「発動できる」（任意）。即時解決せず確認付きで待ち行列へ。
                    # 複数枚（ダブルアタック等）でも確認/解決が中断を跨いで消失しない。
                    if trigger_ability:
                        gm._enqueue_trigger(target_owner, trigger_ability, life_card, optional=True)
                else: gm.winner = attacker_owner.name; break
    else:
        if attacker_pwr >= target_pwr:
            if gm._active_protection(target, ("BATTLE_KO",), attacker=attacker):
                pass
            else:
                repl = gm._find_replacement(target, ("BATTLE_KO",))
                if repl is not None and getattr(repl[3], "is_optional", False):
                    # 任意のバトルKO置換（「代わりに〜してもよい/できる」OP10-034 等）は、
                    # 被KO側に「代わりの効果を使うか」を確認するため戦闘を中断する。
                    # accept→置換実行（本来のKOをスキップ）、decline→本来のKOを実行。
                    # どちらの分岐も resume 時に _finish_attack で戦闘後処理を行う。
                    gm._suspend_for_battle_ko_replacement(target, target_owner, life_lost)
                    return
                elif repl is not None:
                    # 任意でない置換は従来どおり即時実行（内側選択はヘッドレス自動解決）。
                    gm._active_replacement(target, ("BATTLE_KO",))
                else:
                    gm.move_card(target, Zone.TRASH, target_owner)
                    gm._resolve_on_ko(target, target_owner, cause="BATTLE")

    gm._finish_attack(target, target_owner, life_lost)

def _finish_attack(gm, target: Card, target_owner: Player, life_lost: int):
    """戦闘解決後の共通後処理。インラインのバトルKO判定からも、任意バトルKO置換の
    確認(CONFIRM_OPTIONAL)からの resume からも呼ばれる。"""
    target.reset_turn_status(keep_don=True); gm.active_battle = None; gm.phase = Phase.MAIN; gm.check_victory()
    gm.continuous.expire("BATTLE_END", gm.turn_count)
    if not gm.winner:
        gm._apply_passive_effects(gm.turn_player)
    # ライフが離れた回数ぶん ON_LIFE_DECREASE を待ち行列へ積み、【トリガー】と共に消化する。
    if life_lost and not gm.winner:
        gm._enqueue_life_decrease(target_owner, life_lost)
    gm._advance_pending_triggers()

def _suspend_for_battle_ko_replacement(gm, target: Card, target_owner: Player, life_lost: int):
    """任意のバトルKO置換を被KO側へ確認するため戦闘を中断する（CONFIRM_OPTIONAL）。
    resume 時: accept→置換実行（KOスキップ）／decline→本来のKO、その後 _finish_attack。
    ヘッドレス/CPU の既定応答(index0=accept)は従来の自動採用と一致する。"""
    gm.active_interaction = {
        "player_id": target_owner.name,
        "action_type": "CONFIRM_OPTIONAL",
        "source_card_name": target.master.name,
        "source_card_uuid": target.uuid,
        "message": f"「{target.master.name}」がバトルでKOされます。代わりの効果を使用しますか？",
        "can_skip": True,
        "continuation": {
            "kind": "BATTLE_KO_REPLACE",
            "source_card_uuid": target.uuid,
            "target_owner_name": target_owner.name,
            "life_lost": life_lost,
        },
    }

def has_blocker(gm, player: Player) -> bool:
    for card in player.field:
        if (not card.is_rest and card.has_keyword("ブロッカー")
                and "BLOCKER_DISABLED" not in card.flags
                and "CANNOT_REST" not in card.timed_flags):
            return True
    return False

def check_victory(gm):
    # デッキアウト: 通常は本人の敗北（相手の勝利）。ただし C10「自分のデッキが0枚に
    # なった場合、敗北する代わりに勝利する」(VICTORY/REPLACE_DECKOUT_LOSS) を持つ場合は
    # 本人の勝利へ置換する（OP03-040 ナミ等）。
    if not gm.p1.deck:
        gm.winner = gm.p1.name if gm._has_deckout_win_replace(gm.p1) else gm.p2.name
    elif not gm.p2.deck:
        gm.winner = gm.p2.name if gm._has_deckout_win_replace(gm.p2) else gm.p1.name

def _has_deckout_win_replace(gm, player) -> bool:
    """player がデッキアウト時の敗北→勝利の置換能力(PASSIVE)を持つか。"""
    units = [player.leader] + list(player.field)
    for card in units:
        if not card or not getattr(card, "master", None) or getattr(card, "negated", False):
            continue
        if getattr(card, "is_effect_negated", False):
            continue
        for ab in card.master.abilities:
            if ab.trigger != TriggerType.PASSIVE:
                continue
            eff = gm._find_action(ab.effect, ActionType.VICTORY)
            if eff is not None and eff.status == "REPLACE_DECKOUT_LOSS":
                return True
    return False

File: rust/opcg_engine/src/rules/battle.rs
//! 戦闘＝Python `opcg_sim/src/core/engine/battle.py`（P2）。
//!
//! アタック宣言 → （ブロッカーが居れば）ブロックステップ → カウンターステップ → 解決 → 後処理。
//! 効果を伴う分岐（ON_ATTACK／ON_OPP_ATTACK／ON_BLOCK／【トリガー】／KO 置換・除去保護・
//! 【カウンター】イベント）は P3 の担当なので、バニラでは走らない経路として
//! `Unimplemented` か「空の待ち行列」で表す。

use crate::journal::{CardBoolField, CardZone, Session, TriggerQueue};
use crate::model::{ActiveBattle, CardIdx, CardType, MasterTable, Phase, Position, Seat, Zone};
use crate::ops;
use crate::state::EngineError;

use super::passive::apply_passive_effects;
use super::{
    card_type, has_flag, has_keyword, has_timed_flag, KW_ATTACK_ACTIVE, KW_BANISH, KW_BLOCKER,
    KW_DOUBLE_ATTACK, KW_RUSH,
};

fn bad(msg: impl Into<String>) -> EngineError {
    EngineError::BadPayload(msg.into())
}

/// Python `has_blocker`（ブロッカーを 1 枚でも構えているか）。
///
/// `get_pending_request` のブロッカー候補列挙とは条件が 1 つ違う（こちらは
/// `BLOCKER_DISABLED` も見る）。Python のまま移す。
pub fn has_blocker(s: &Session, seat: Seat) -> bool {
    let st = s.state();
    st.player(seat).field.iter().any(|c| {
        !st.card(*c).is_rest
            && has_keyword(st, *c, KW_BLOCKER)
            && !st.card(*c).flags.iter().any(|f| f == "BLOCKER_DISABLED")
            && !has_timed_flag(st, *c, "CANNOT_REST")
    })
}

/// Python `declare_attack`。検証は全て同じ順序・同じメッセージで行う。
pub fn declare_attack(
    s: &mut Session,
    masters: &MasterTable,
    attacker: CardIdx,
    target: CardIdx,
) -> Result<(), EngineError> {
    let attacker_owner = ops::find_card_location(s.state(), attacker)
        .map(|(seat, _)| seat)
        .ok_or_else(|| bad("declare_attack: アタッカーの所在が見つかりません。"))?;
    let target_owner = ops::find_card_location(s.state(), target)
        .map(|(seat, _)| seat)
        .ok_or_else(|| bad("declare_attack: 攻撃対象の所在が見つかりません。"))?;
    super::actions::validate_action(s, attacker_owner, "MAIN_ACTION")?;

    // 先攻・後攻ともに「自分の最初のターン」（turn_count <= 2）はアタックできない。
    if s.state().turn_count <= 2 {
        return Err(bad("最初のターンはアタックできません。"));
    }
    if has_flag(s.state(), attacker, "ATTACK_DISABLE") {
        return Err(bad("このカードは効果によりアタックできません。"));
    }
    if has_timed_flag(s.state(), attacker, "CANNOT_REST") {
        return Err(bad(
            "このカードは効果によりレストにできないためアタックできません。",
        ));
    }
    if s.state().card(attacker).is_rest {
        return Err(bad(
            "アタックするカードはアクティブ状態でなければなりません。",
        ));
    }
    // 召喚酔い（登場したターンのキャラ。速攻を持てば可。リーダーは is_newly_played=false）。
    if card_type(s.state(), masters, attacker) == CardType::Character
        && s.state().card(attacker).is_newly_played
        && !has_keyword(s.state(), attacker, KW_RUSH)
    {
        return Err(bad(
            "登場したターンのキャラクターは攻撃できません（速攻を除く）。",
        ));
    }
    if card_type(s.state(), masters, target) == CardType::Leader
        && super::active_restriction(s.state(), attacker_owner, "CANNOT_ATTACK_LEADER").is_some()
    {
        return Err(bad(
            "効果により、このターンはリーダーにアタックできません。",
        ));
    }
    if card_type(s.state(), masters, target) == CardType::Character
        && !s.state().card(target).is_rest
        && !has_keyword(s.state(), attacker, KW_ATTACK_ACTIVE)
    {
        return Err(bad("レスト状態のキャラクターのみ攻撃可能です。"));
    }
    // アタック税（ATTACK_TAX_DISCARD_N）は効果由来のフラグ＝バニラでは立たない（P3）。
    if s.state()
        .card(attacker)
        .flags
        .iter()
        .chain(s.state().card(attacker).timed_flags.iter())
        .any(|f| f.starts_with("ATTACK_TAX_DISCARD_"))
    {
        return Err(EngineError::Unimplemented(
            "declare_attack: アタック税（ATTACK_TAX_DISCARD_N）は効果由来＝P3".into(),
        ));
    }

    ops::set_rest(s, attacker, true);
    s.edit().set_active_battle(Some(ActiveBattle {
        attacker,
        target,
        attacker_owner,
        target_owner,
        counter_buff: 0,
    }));
    // Python はここで ON_ATTACK / ON_REST / ON_OPP_ATTACK を `_battle_triggers` へ積む。
    // バニラは abilities が無いので**空**（待ち行列だけ同じ形で持つ）。
    s.edit().set_trigger_queue(TriggerQueue::Battle, Vec::new());
    advance_battle_triggers(s, masters);
    Ok(())
}

/// Python `_advance_battle_triggers`: 積んだトリガーを解決し、終わったら防御フェイズへ。
pub fn advance_battle_triggers(s: &mut Session, _masters: &MasterTable) {
    let Some(battle) = s.state().active_battle.clone() else {
        s.edit().set_trigger_queue(TriggerQueue::Battle, Vec::new());
        return;
    };
    // バニラでは `battle_triggers` は常に空（効果解決は P3）。
    debug_assert!(s.state().battle_triggers.is_empty());
    let phase = if has_blocker(s, battle.target_owner) {
        Phase::BlockStep
    } else {
        Phase::BattleCounter
    };
    s.edit().set_phase(phase);
}

/// Python `handle_block`。`blocker=None`（ブロックしない）はカウンターステップへ進むだけ。
pub fn handle_block(
    s: &mut Session,
    blocker: Option<CardIdx>,
) -> Result<(), EngineError> {
    let Some(battle) = s.state().active_battle.clone() else {
        return Ok(());
    };
    super::actions::validate_action(s, battle.target_owner, "SELECT_BLOCKER")?;
    if let Some(blocker) = blocker {
        s.edit()
            .set_card_bool(blocker, CardBoolField::IsRest, true);
        let mut updated = battle.clone();
        updated.target = blocker;
        s.edit().set_active_battle(Some(updated));
        // Python: ON_BLOCK（【ブロック時】）の解決＝P3。バニラでは能力が無い。
    }
    s.edit().set_phase(Phase::BattleCounter);
    Ok(())
}

/// Python `apply_counter`。`counter_card=None` は「カウンターしない」＝解決へ。
pub fn apply_counter(
    s: &mut Session,
    masters: &MasterTable,
    seat: Seat,
    counter_card: Option<CardIdx>,
) -> Result<(), EngineError> {
    if s.state().active_battle.is_none() {
        return Ok(());
    }
    let Some(counter_card) = counter_card else {
        return resolve_attack(s, masters);
    };
    super::actions::validate_action(s, seat, "SELECT_COUNTER")?;
    if card_type(s.state(), masters, counter_card) == CardType::Event {
        // 【カウンター】イベントは効果解決（P3）が要る。バニラでは候補に出ない
        // （`counter_candidates` はカウンター値を持つ手札だけを出す）。
        return Err(EngineError::Unimplemented(
            "apply_counter: 【カウンター】イベントの効果解決は P3".into(),
        ));
    }
    let value = super::current_counter(s.state(), masters, counter_card);
    let mut battle = s.state().active_battle.clone().expect("checked above");
    battle.counter_buff += value;
    s.edit().set_active_battle(Some(battle));
    ops::move_card(s, masters, counter_card, Zone::Trash, seat, Position::Bottom)?;
    Ok(())
}

/// Python `resolve_attack`。
pub fn resolve_attack(s: &mut Session, masters: &MasterTable) -> Result<(), EngineError> {
    let Some(battle) = s.state().active_battle.clone() else {
        return Ok(());
    };
    let ActiveBattle {
        attacker,
        target,
        attacker_owner,
        target_owner,
        counter_buff,
    } = battle;

    let turn_player = s.state().turn_player;
    let is_my_turn = attacker_owner == turn_player;
    let is_target_turn = target_owner == turn_player;
    let attacker_pwr = {
        let c = s.state().card(attacker);
        c.get_power(masters.get(c.master), is_my_turn)
    };
    let target_pwr = {
        let c = s.state().card(target);
        c.get_power(masters.get(c.master), is_target_turn)
    } + counter_buff;

    let mut life_lost = 0;
    if s.state().player(target_owner).leader == Some(target) {
        if attacker_pwr >= target_pwr {
            let damage = if has_keyword(s.state(), attacker, KW_DOUBLE_ATTACK) {
                2
            } else {
                1
            };
            let banish = has_keyword(s.state(), attacker, KW_BANISH);
            for _ in 0..damage {
                if s.state().player(target_owner).life.is_empty() {
                    s.edit().set_winner(Some(attacker_owner));
                    break;
                }
                // Python は `life.pop(0)` してから move_card する（＝move_card 側から
                // ライフ離脱には見えない。ON_LIFE_DECREASE は下でまとめて積む）。
                let life_card = s
                    .edit()
                    .card_zone_remove_at(target_owner, CardZone::Life, 0);
                let dest = if banish { Zone::Trash } else { Zone::Hand };
                ops::move_card(s, masters, life_card, dest, target_owner, Position::Bottom)?;
                life_lost += 1;
                // Python: 公開した【トリガー】は確認付きで待ち行列へ（P3）。バニラでは無い。
            }
        }
    } else if attacker_pwr >= target_pwr {
        // 除去保護（PREVENT_LEAVE/BATTLE_KO）・KO 置換は効果＝P3。バニラでは素の KO。
        ops::move_card(s, masters, target, Zone::Trash, target_owner, Position::Bottom)?;
        resolve_on_ko(s, target_owner);
    }

    finish_attack(s, masters, target, life_lost);
    Ok(())
}

/// Python `_resolve_on_ko` のうち P2 が担う部分＝ターン内イベントの記録。
/// 【KO時】誘発と第三者 KO リスナーは効果（P3）。
fn resolve_on_ko(s: &mut Session, owner: Seat) {
    let name = format!("CHAR_KOED_{}", owner.name());
    ops::record_turn_event(s, &name, 1);
}

/// Python `_finish_attack`（戦闘解決後の共通後処理）。
fn finish_attack(s: &mut Session, masters: &MasterTable, target: CardIdx, life_lost: i32) {
    ops::reset_turn_status(s, masters, target, true, false);
    s.edit().set_active_battle(None);
    s.edit().set_phase(Phase::Main);
    check_victory(s);
    // Python: continuous.expire("BATTLE_END", turn_count)＝継続効果の失効（P3）。
    if s.state().winner.is_none() {
        let tp = s.state().turn_player;
        apply_passive_effects(s, masters, tp);
    }
    // Python: ライフが離れた回数ぶん ON_LIFE_DECREASE を積む（P3。バニラでは空）。
    let _ = life_lost;
}

/// Python `check_victory`（デッキアウト。勝敗の置換 REPLACE_DECKOUT_LOSS は効果＝P3）。
pub fn check_victory(s: &mut Session) {
    if s.state().player(Seat::P1).deck.is_empty() {
        s.edit().set_winner(Some(Seat::P2));
    } else if s.state().player(Seat::P2).deck.is_empty() {
        s.edit().set_winner(Some(Seat::P1));
    }
}
